### PR TITLE
refactor(factory-reset): new interface to provide filepaths to restore

### PR DESCRIPTION
**Pull request by mlilien**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -107,9 +107,9 @@ Result:
 }
 
 
-The supported reset `type` and the documentation in general can be found in the [meta-omnect layer](https://github.com/omnect/meta-omnect#factory-reset).
+The supported reset `mode` and the documentation in general can be found in the [meta-omnect layer](https://github.com/omnect/meta-omnect#factory-reset).
 
-The **optional** `restore_settings` array can be used to define user settings that must be restored after wiping device storage. Currently only `wifi` settings in `wpa_supplicant.conf` can be restored.
+The **optional** `preserve` array can be used to define system resp. user settings that must be restored after wiping device storage. Supported are keys from [here](https://github.com/omnect/meta-omnect/blob/main/recipes-omnect/omnect-device-service/omnect-device-service/factory-reset.json) and the key "applications" if there is a [custom configuration file](https://github.com/omnect/meta-omnect#factory-resetTODO) in `/etc/omnect/factory-reset.d/`.
```
Comment (mlilien) [inline README.md:112]: ToDo: correct link as soon custom configuration is documented in meta-omnect

File: src/twin/factory_reset.rs [hunk 2]
```diff
@@ -99,51 +151,28 @@ impl FactoryReset {
...
         }
+
+        bootloader_env::set("factory-reset", &in_json.to_string())?;
+        self.report_factory_reset_status("in_progress").await?;
+        systemd::reboot().await?;
+        Ok(None)
     }
 
     async fn report_factory_reset_status(&self, status: &str) -> Result<()> {
+        // ToDo why is that not the same format as for the cloud?
```
Comment (mlilien) [inline src/twin/factory_reset.rs:175]: @JanZachmann

Comment (JanZachmann) [reply src/twin/factory_reset.rs:175]: No special reason and we should make it equal. Pls create a task for omnect-ui to adapt to the new structure.

File: src/twin/factory_reset.rs [hunk 3]
```diff
@@ -170,14 +199,28 @@ impl FactoryReset {
...
+                "{}:{}",
+                status.to_string().trim_matches('"'),
+                error.to_string().trim_matches('"')
             );
-        };
+        }
 
-        match factory_reset_status.trim_end() {
+        debug!("factory_reset_status: {factory_reset_status}");
+        // ToDo more stati
```
Comment (JanZachmann) [inline src/twin/factory_reset.rs:222]: Imho we should do this in a new PBI "omnet-portal: adapt to factory-resetv2 + fsck status"

File: src/twin/factory_reset.rs [hunk 4]
```diff
@@ -72,25 +92,57 @@ impl Feature for FactoryReset {
...
+    {
+        let key_value_map: HashMap<String, serde_json::Value> = serde_json::from_reader(BufReader::new(
+            File::open(factory_reset_config_path!()).context("open factory-reset.json")?,
+        ))
+        .context("parsing factory reset config")?;
+
+        let mut keys: Vec<String> = key_value_map.into_keys().collect();
+        if read_dir(factory_reset_custom_config_dir_path!()).context("read factory-reset.d")?.next().is_some(){
+            keys.push(String::from("applications"))
+        }
```
[the remark this answers is not in the document]
Comment (mlilien) [reply src/twin/factory_reset.rs:114]: why? performance related? can you point me to more informations about that topic?
and with what value?

File: src/twin/factory_reset.rs [hunk 5]
```diff
@@ -72,25 +92,57 @@ impl Feature for FactoryReset {
...
+
+        #[cfg(feature = "mock")]
+        keys.sort_by(|a, b| a.to_lowercase().cmp(&b.to_lowercase()));
```
[the remark this answers is not in the document]
Comment (mlilien) [reply src/twin/factory_reset.rs:117]: yes, without sorting expectations to the twin mock reported properties sometime failed

File: src/twin/factory_reset.rs [hunk 6]
```diff
@@ -99,51 +151,28 @@ impl FactoryReset {
...
+        if in_json["mode"].as_u64().is_none() {
+            anyhow::bail!("reset mode missing or not supported");
+        }
+        let preserve = in_json["preserve"].as_array();
+        if  preserve.is_some() {
+            let keys=FactoryReset::factory_reset_keys()?;
+            for topic in preserve.unwrap().iter() {
+                let topic = String::from(topic.to_string().trim_matches('"'));
+                if ! keys.contains(&topic) {
+                    anyhow::bail!("unknown preserve topic received: {topic}");
```
[the remark this answers is not in the document]
Comment (mlilien) [reply src/twin/factory_reset.rs:163]: I doubt i can bail in the closure ... but'll give it a try.

File: src/twin/factory_reset.rs (hunk 2 above)
Comment (mlilien) [reply src/twin/factory_reset.rs:175]: task imho already exists, created by you.

File: src/twin/factory_reset.rs [hunk 7]
```diff
@@ -99,51 +151,28 @@ impl FactoryReset {
...
-                            paths.push(path.clone());
-                        } else {
-                            anyhow::bail!("invalid restore path received: {path}");
-                        }
-                    } else {
-                        anyhow::bail!("unknown restore setting received: {s}");
-                    }
+        if in_json["mode"].as_u64().is_none() {
+            anyhow::bail!("reset mode missing or not supported");
+        }
```
[the remark this answers is not in the document]
Comment (mlilien) [reply src/twin/factory_reset.rs:156]: correct. i'll add a test to the integration test though.

File: src/twin/factory_reset.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/factory_reset.rs:114]: overlooked that you return just the keys. thus it's fine...

File: src/twin/factory_reset.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/factory_reset.rs:163]: just bail in the `if` body without logging the topic or logging entire topic vector?
```
  let in_json = Some(vec!["item1", "item2"]);
  if let Some(preserve) = in_json {
      let keys=vec!["item1", "item3"];
      if !preserve.iter().all(|item| keys.contains(&item.trim_matches('"'))) {
            anyhow::bail!("unknown preserve topic received");
      }
  }
```

File: src/web_service.rs [hunk 8]
```diff
@@ -101,7 +102,7 @@ impl WebService {
         let srv = HttpServer::new(move || {
             App::new()
                 .app_data(web::Data::new(tx_request.clone()))
-                .route("/factory-reset/v1", web::post().to(Self::factory_reset))
+                .route("/factory-reset/v2", web::post().to(Self::factory_reset))
```
Comment (JanZachmann) [inline src/web_service.rs:105]: revert

File: src/twin/factory_reset.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mlilien) [reply src/twin/factory_reset.rs:163]: i leave it as is, since i want to inform about the topic in question.

File: src/web_service.rs (hunk 8 above)
Comment (mlilien) [reply src/web_service.rs]: done

**Review by JanZachmann — APPROVED** [review]

**Pull request merged**